### ISSUE TITLE
Adding a node

### ISSUE THREAD

**Issue by moserhans** [OndrejHome/ansible.pcs-modules-2#4]
I can create a cluster with your helpful role, thank you.
Nevertheless if a add a new host to the role/to the cluster, this fails.
The corosync.conf is not changed and not copied to the new node.

I think the role does not handle the adding of a new node, is that right?

**Comment by OndrejHome** [comment]
Hi @moserhans. That is correct. modules are not handling adding nor removing the nodes. I would have to check if making this would be easy to do. I'm unsure about the possibility for removing. Would you interested in contributing the change for adding the node to modules or at least test it in case this is added to modules here?

**Comment by moserhans** [comment]
Thank you.
I will see what I can do to help.

**Comment by OndrejHome** [comment]
Having a look at how the code of this looks it may not be so simple to "just add node". Currently code only tries the 'pcs cluster setup' when cluster configuration files are not detected. This is very basic stuff that will not be correct in some scenarios. Also the 'pcs_cluster' module expects that it will run only on one node and add the other nodes.

I would have to think if there is way how to make this to work properly with 'pcs node add' and if that is possible at all. Maybe parsing the '/etc/corosync/corosync.conf' can be a way to go (like detecting the node names from it and comparing with given argument. 

If you have ideal on what would be a preferred way on how would you like to use this module let me know. Also if you have ideas on how to make this in code feel free to comment. I would see if something comes on my mind during week and update here any new ideas.

**Comment by moserhans** [comment]
Thank you.
I incountered the "add node" problem in my test phase, when I addes more, "fresh" nodes to have a clean base. For what we have in mind to use your very usefull module, this is not essential.

**Comment by OndrejHome** [comment]
Thank you for feedback on use case. I will keep this Issue open as it  looms interesting to me, but I'm not sure when I will have time to have into this. If this would be possible I would like to be able to use this together with my other role OndrejHome.ha-cluster-pacemaker to add another node to cluster - that role covers installation and enablement of cluster which makes it ideal candidate for testing adding the node. For now I will keep it assigned only to me a update it here once there is a progress.

**Comment by OndrejHome** [comment]
So revisiting this thing after longer time as it seems that there are also some other people interested in this feature that contacted me via email.

I have started drafting a first idea on how this could possibly work in way that would make this least painful for user to do. So far initial idea is to use "action plugin" to do decision making.

Idea is that before actual 'pcs_cluster' module will do some work, the 'pcs_cluster' action plugin will determine on which nodes the 'pcs_cluster' module should be run and if any parameters for it needs to be changed. Before the 'pcs_cluster' module action plugin can decide on what to do it will need some data - this data would be collected by facts plugin ('pacemaker_cluster_detect') and stored in the hostvars of nodes (this still needs to be written, currently simulated manually via inventory file). 

There is so far one functional decision that I have made - at one time it would be possible to "only add" or "only remove" nodes. This is done to make it possible for end user to decide what should be done first.

Adding and removing nodes must be explicitly allowed via 'pcs cluster' attribute, without that no node will be removed or added.

I have started a separate branch for this and initial code so far only makes decisions on what to do.
Still a very dirty version that is not functional, but gives idea for proof of concept.

https://github.com/OndrejHome/ansible.pcs-modules-2/tree/pcs_cluster_add_node

This would need quite a time to implement, but I have at least now an idea on how to proceed. Once I have something that can be tested I will update this Issue with details.
